File: app/src/main/java/at/bitfire/davdroid/resource/LocalGroup.java
/*
 * Copyright © 2013 – 2015 Ricki Hirner (bitfire web engineering).
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the GNU Public License v3.0
 * which accompanies this distribution, and is available at
 * http://www.gnu.org/licenses/gpl.html
 */

package at.bitfire.davdroid.resource;

import android.content.ContentProviderOperation;
import android.content.ContentUris;
import android.content.ContentValues;
import android.database.Cursor;
import android.os.Parcel;
import android.os.RemoteException;
import android.provider.ContactsContract;
import android.provider.ContactsContract.CommonDataKinds.GroupMembership;
import android.provider.ContactsContract.Groups;
import android.provider.ContactsContract.RawContacts;
import android.provider.ContactsContract.RawContacts.Data;

import org.apache.commons.lang3.ArrayUtils;

import java.io.FileNotFoundException;
import java.util.LinkedList;
import java.util.List;
import java.util.UUID;
import java.util.logging.Level;

import at.bitfire.dav4android.Constants;
import at.bitfire.vcard4android.AndroidAddressBook;
import at.bitfire.vcard4android.AndroidGroup;
import at.bitfire.vcard4android.AndroidGroupFactory;
import at.bitfire.vcard4android.BatchOperation;
import at.bitfire.vcard4android.CachedGroupMembership;
import at.bitfire.vcard4android.Contact;
import at.bitfire.vcard4android.ContactsStorageException;
import lombok.Cleanup;
import lombok.ToString;

@ToString(callSuper=true)
public class LocalGroup extends AndroidGroup implements LocalResource {
    /** marshalled list of member UIDs, as sent by server */
    public static final String COLUMN_PENDING_MEMBERS = Groups.SYNC3;

    public LocalGroup(AndroidAddressBook addressBook, long id, String fileName, String eTag) {
        super(addressBook, id, fileName, eTag);
    }

    public LocalGroup(AndroidAddressBook addressBook, Contact contact, String fileName, String eTag) {
        super(addressBook, contact, fileName, eTag);
    }


    @Override
    public void clearDirty(String eTag) throws ContactsStorageException {
        assertID();

        ContentValues values = new ContentValues(2);
        values.put(Groups.DIRTY, 0);
        values.put(COLUMN_ETAG, this.eTag = eTag);
        update(values);

        // update cached group memberships
        BatchOperation batch = new BatchOperation(addressBook.provider);

        // delete cached group memberships
        batch.enqueue(new BatchOperation.Operation(
                ContentProviderOperation.newDelete(addressBook.syncAdapterURI(ContactsContract.Data.CONTENT_URI))
                        .withSelection(
                                CachedGroupMembership.MIMETYPE + "=? AND " + CachedGroupMembership.GROUP_ID + "=?",
                                new String[] { CachedGroupMembership.CONTENT_ITEM_TYPE, String.valueOf(id) }
                        )
        ));

        // insert updated cached group memberships
        for (long member : getMembers())
            batch.enqueue(new BatchOperation.Operation(
                    ContentProviderOperation.newInsert(addressBook.syncAdapterURI(ContactsContract.Data.CONTENT_URI))
                            .withValue(CachedGroupMembership.MIMETYPE, CachedGroupMembership.CONTENT_ITEM_TYPE)
                            .withValue(CachedGroupMembership.RAW_CONTACT_ID, member)
                            .withValue(CachedGroupMembership.GROUP_ID, id)
                            .withYieldAllowed(true)
            ));

        batch.commit();
    }

    @Override
    public void prepareForUpload() throws ContactsStorageException {
        final String uid = UUID.randomUUID().toString();
        final String newFileName = uid + ".vcf";

        ContentValues values = new ContentValues(2);
        values.put(COLUMN_FILENAME, newFileName);
        values.put(COLUMN_UID, uid);
        update(values);

        fileName = newFileName;
    }

    @Override
    protected ContentValues contentValues() {
        ContentValues values = super.contentValues();

        @Cleanup("recycle") Parcel members = Parcel.obtain();
        members.writeStringList(contact.members);
        values.put(COLUMN_PENDING_MEMBERS, members.marshall());

        return values;
    }


    /**
     * Marks all members of the current group as dirty.
     */
    public void markMembersDirty() throws ContactsStorageException {
        assertID();
        BatchOperation batch = new BatchOperation(addressBook.provider);

        for (long member : getMembers())
            batch.enqueue(new BatchOperation.Operation(
                    ContentProviderOperation.newUpdate(addressBook.syncAdapterURI(ContentUris.withAppendedId(RawContacts.CONTENT_URI, member)))
                            .withValue(RawContacts.DIRTY, 1)
                            .withYieldAllowed(true)
            ));

        batch.commit();
    }

    /**
     * Processes all groups with non-null {@link #COLUMN_PENDING_MEMBERS}: the pending memberships
     * are (if possible) applied, keeping cached memberships in sync.
     * @param addressBook    address book to take groups from
     * @throws ContactsStorageException on contact provider errors
     */
    public static void applyPendingMemberships(LocalAddressBook addressBook) throws ContactsStorageException {
        try {
            @Cleanup Cursor cursor = addressBook.provider.query(
                    addressBook.syncAdapterURI(Groups.CONTENT_URI),
                    new String[] { Groups._ID, COLUMN_PENDING_MEMBERS },
                    COLUMN_PENDING_MEMBERS + " IS NOT NULL", new String[] {},
                    null
            );

            BatchOperation batch = new BatchOperation(addressBook.provider);
            while (cursor != null && cursor.moveToNext()) {
                long id = cursor.getLong(0);
                Constants.log.fine("Assigning members to group " + id);

                // delete all memberships and cached memberships for this group
                batch.enqueue(new BatchOperation.Operation(
                        ContentProviderOperation.newDelete(addressBook.syncAdapterURI(ContactsContract.Data.CONTENT_URI))
                                .withSelection(
                                        "(" + GroupMembership.MIMETYPE + "=? AND " + GroupMembership.GROUP_ROW_ID + "=?) OR (" +
                                              CachedGroupMembership.MIMETYPE + "=? AND " + CachedGroupMembership.GROUP_ID + "=?)",
                                        new String[] { GroupMembership.CONTENT_ITEM_TYPE, String.valueOf(id), CachedGroupMembership.CONTENT_ITEM_TYPE, String.valueOf(id) })
                                .withYieldAllowed(true)
                ));

                // extract list of member UIDs
                List<String> members = new LinkedList<>();
                byte[] raw = cursor.getBlob(1);
                @Cleanup("recycle") Parcel parcel = Parcel.obtain();
                parcel.unmarshall(raw, 0, raw.length);
                parcel.setDataPosition(0);
                parcel.readStringList(members);

                // insert memberships
                for (String uid : members) {
                    Constants.log.fine("Assigning member: " + uid);
                    try {
                        LocalContact member = addressBook.findContactByUID(uid);
                        member.addToGroup(batch, id);
                    } catch(FileNotFoundException e) {
                        Constants.log.log(Level.WARNING, "Group member not found: " + uid, e);
                    }
                }

                // remove pending memberships
                batch.enqueue(new BatchOperation.Operation(
                        ContentProviderOperation.newUpdate(addressBook.syncAdapterURI(ContentUris.withAppendedId(Groups.CONTENT_URI, id)))
                                .withValue(COLUMN_PENDING_MEMBERS, null)
                                .withYieldAllowed(true)
                ));

                batch.commit();
            }
        } catch(RemoteException e) {
            throw new ContactsStorageException("Couldn't get pending memberships", e);
        }
    }


    // helpers

    private void assertID() {
        if (id == null)
            throw new IllegalStateException("Group has not been saved yet");
    }

    /**
     * Lists all members of this group.
     * @return list of all members' raw contact IDs
     * @throws ContactsStorageException on contact provider errors
     */
    protected long[] getMembers() throws ContactsStorageException {
        assertID();
        List<Long> members = new LinkedList<>();
        try {
            @Cleanup Cursor cursor = addressBook.provider.query(
                    addressBook.syncAdapterURI(ContactsContract.Data.CONTENT_URI),
                    new String[] { Data.RAW_CONTACT_ID },
                    GroupMembership.MIMETYPE + "=? AND " + GroupMembership.GROUP_ROW_ID + "=?",
                    new String[] { GroupMembership.CONTENT_ITEM_TYPE, String.valueOf(id) },
                    null
            );
            while (cursor != null && cursor.moveToNext())
                members.add(cursor.getLong(0));
        } catch(RemoteException e) {
            throw new ContactsStorageException("Couldn't list group members", e);
        }
        return ArrayUtils.toPrimitive(members.toArray(new Long[members.size()]));
    }


    // factory

    static class Factory extends AndroidGroupFactory {
        static final Factory INSTANCE = new Factory();

        @Override
        public LocalGroup newInstance(AndroidAddressBook addressBook, long id, String fileName, String eTag) {
            return new LocalGroup(addressBook, id, fileName, eTag);
        }

        @Override
        public LocalGroup newInstance(AndroidAddressBook addressBook, Contact contact, String fileName, String eTag) {
            return new LocalGroup(addressBook, contact, fileName, eTag);
        }

        @Override
        public LocalGroup[] newArray(int size) {
            return new LocalGroup[size];
        }

    }

}

File: app/src/main/kotlin/at/bitfire/davdroid/resource/LocalGroup.kt
/***************************************************************************************************
 * Copyright © All Contributors. See LICENSE and AUTHORS in the root directory for details.
 **************************************************************************************************/

package at.bitfire.davdroid.resource

import android.content.ContentUris
import android.content.ContentValues
import android.net.Uri
import android.os.Build
import android.os.RemoteException
import android.provider.ContactsContract
import android.provider.ContactsContract.CommonDataKinds.GroupMembership
import android.provider.ContactsContract.Groups
import android.provider.ContactsContract.RawContacts
import android.provider.ContactsContract.RawContacts.Data
import at.bitfire.davdroid.log.Logger
import at.bitfire.vcard4android.*
import org.apache.commons.lang3.StringUtils
import java.util.*

class LocalGroup: AndroidGroup, LocalAddress {

    companion object {

        const val COLUMN_FLAGS = Groups.SYNC4

        /** List of member UIDs, as sent by server. This list will be used to establish
         *  the group memberships when all groups and contacts have been synchronized.
         *  Use [PendingMemberships] to create/read the list. */
        const val COLUMN_PENDING_MEMBERS = Groups.SYNC3

        /**
         * Processes all groups with non-null [COLUMN_PENDING_MEMBERS]: the pending memberships
         * are applied (if possible) to keep cached memberships in sync.
         *
         * @param addressBook    address book to take groups from
         */
        fun applyPendingMemberships(addressBook: LocalAddressBook) {
            Logger.log.info("Assigning memberships of contact groups")

            addressBook.allGroups { group ->
                val groupId = group.id!!
                val pendingMemberUids = group.pendingMemberships.toMutableSet()
                val batch = BatchOperation(addressBook.provider!!)

                // required for workaround for Android 7 which sets DIRTY flag when only meta-data is changed
                val changeContactIDs = HashSet<Long>()

                // process members which are currently in this group, but shouldn't be
                for (currentMemberId in addressBook.getContactIdsByGroupMembership(groupId)) {
                    val uid = addressBook.getContactUidFromId(currentMemberId) ?: continue

                    if (!pendingMemberUids.contains(uid)) {
                        Logger.log.fine("$currentMemberId removed from group $groupId; removing group membership")
                        val currentMember = addressBook.findContactById(currentMemberId)
                        currentMember.removeGroupMemberships(batch)

                        // Android 7 hack
                        changeContactIDs += currentMemberId
                    }

                    // UID is processed, remove from pendingMembers
                    pendingMemberUids -= uid
                }
                // now pendingMemberUids contains all UIDs which are not assigned yet

                // process members which should be in this group, but aren't
                for (missingMemberUid in pendingMemberUids) {
                    val missingMember = addressBook.findContactByUid(missingMemberUid)
                    if (missingMember == null) {
                        Logger.log.warning("Group $groupId has member $missingMemberUid which is not found in the address book; ignoring")
                        continue
                    }

                    Logger.log.fine("Assigning member $missingMember to group $groupId")
                    missingMember.addToGroup(batch, groupId)

                    // Android 7 hack
                    changeContactIDs += missingMember.id!!
                }

                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N && Build.VERSION.SDK_INT < Build.VERSION_CODES.O)
                    // workaround for Android 7 which sets DIRTY flag when only meta-data is changed
                    changeContactIDs
                            .map { addressBook.findContactById(it) }
                            .forEach { it.updateHashCode(batch) }

                batch.commit()
            }
        }

    }


    override var scheduleTag: String?
        get() = null
        set(value) = throw NotImplementedError()

    override var flags: Int = 0

    var pendingMemberships = setOf<String>()


    constructor(addressBook: AndroidAddressBook<out AndroidContact, LocalGroup>, values: ContentValues) : super(addressBook, values) {
        flags = values.getAsInteger(COLUMN_FLAGS) ?: 0
        values.getAsString(COLUMN_PENDING_MEMBERS)?.let { members ->
            pendingMemberships = PendingMemberships.fromString(members).uids
        }
    }

    constructor(addressBook: AndroidAddressBook<out AndroidContact, LocalGroup>, contact: Contact, fileName: String?, eTag: String?, flags: Int)
        : super(addressBook, contact, fileName, eTag) {
        this.flags = flags
    }


    override fun contentValues(): ContentValues  {
        val values = super.contentValues()
        values.put(COLUMN_FLAGS, flags)
        values.put(COLUMN_PENDING_MEMBERS, PendingMemberships(getContact().members).toString())
        return values
    }


    override fun prepareForUpload(): String {
        var uid: String? = null
        addressBook.provider!!.query(groupSyncUri(), arrayOf(AndroidContact.COLUMN_UID), null, null, null)?.use { cursor ->
            if (cursor.moveToNext())
                uid = StringUtils.trimToNull(cursor.getString(0))
        }

        if (uid == null) {
            // generate new UID
            uid = UUID.randomUUID().toString()

            val values = ContentValues(1)
            values.put(AndroidContact.COLUMN_UID, uid)
            addressBook.provider!!.update(groupSyncUri(), values, null, null)

            _contact?.uid = uid
        }

        return "$uid.vcf"
    }

    override fun clearDirty(fileName: String?, eTag: String?, scheduleTag: String?) {
        if (scheduleTag != null)
            throw IllegalArgumentException("Contact groups must not have a Schedule-Tag")
        val id = requireNotNull(id)

        val values = ContentValues(3)
        if (fileName != null)
            values.put(COLUMN_FILENAME, fileName)
        values.putNull(COLUMN_ETAG)     // don't save changed ETag but null, so that the group is downloaded again, so that pendingMembers is updated
        values.put(Groups.DIRTY, 0)
        update(values)

        if (fileName != null)
            this.fileName = fileName
        this.eTag = null

        // update cached group memberships
        val batch = BatchOperation(addressBook.provider!!)

        // delete old cached group memberships
        batch.enqueue(BatchOperation.CpoBuilder
                .newDelete(addressBook.syncAdapterURI(ContactsContract.Data.CONTENT_URI))
                .withSelection(
                        CachedGroupMembership.MIMETYPE + "=? AND " + CachedGroupMembership.GROUP_ID + "=?",
                        arrayOf(CachedGroupMembership.CONTENT_ITEM_TYPE, id.toString())
                ))

        // insert updated cached group memberships
        for (member in getMembers())
            batch.enqueue(BatchOperation.CpoBuilder
                    .newInsert(addressBook.syncAdapterURI(ContactsContract.Data.CONTENT_URI))
                    .withValue(CachedGroupMembership.MIMETYPE, CachedGroupMembership.CONTENT_ITEM_TYPE)
                    .withValue(CachedGroupMembership.RAW_CONTACT_ID, member)
                    .withValue(CachedGroupMembership.GROUP_ID, id))

        batch.commit()
    }

    /**
     * Marks all members of the current group as dirty.
     */
    fun markMembersDirty() {
        val batch = BatchOperation(addressBook.provider!!)

        for (member in getMembers())
            batch.enqueue(BatchOperation.CpoBuilder
                    .newUpdate(addressBook.syncAdapterURI(ContentUris.withAppendedId(RawContacts.CONTENT_URI, member)))
                    .withValue(RawContacts.DIRTY, 1))

        batch.commit()
    }

    override fun resetDeleted() {
        val values = ContentValues(1)
        values.put(Groups.DELETED, 0)
        addressBook.provider!!.update(groupSyncUri(), values, null, null)
    }

    override fun updateFlags(flags: Int) {
        val values = ContentValues(1)
        values.put(COLUMN_FLAGS, flags)
        addressBook.provider!!.update(groupSyncUri(), values, null, null)

        this.flags = flags
    }


    // helpers

    private fun groupSyncUri(): Uri {
        val id = requireNotNull(id)
        return ContentUris.withAppendedId(addressBook.groupsSyncUri(), id)
    }

    /**
     * Lists all members of this group.
     * @return list of all members' raw contact IDs
     * @throws RemoteException on contact provider errors
     */
    internal fun getMembers(): List<Long> {
        val id = requireNotNull(id)
        val members = LinkedList<Long>()
        addressBook.provider!!.query(
                addressBook.syncAdapterURI(ContactsContract.Data.CONTENT_URI),
                arrayOf(Data.RAW_CONTACT_ID),
                "${GroupMembership.MIMETYPE}=? AND ${GroupMembership.GROUP_ROW_ID}=?",
                arrayOf(GroupMembership.CONTENT_ITEM_TYPE, id.toString()),
                null
        )?.use { cursor ->
            while (cursor.moveToNext())
                members += cursor.getLong(0)
        }
        return members
    }


    // helper class for COLUMN_PENDING_MEMBERSHIPS blob

    class PendingMemberships(
        /** list of member UIDs that shall be assigned **/
        val uids: Set<String>
    ) {

        companion object {
            const val SEPARATOR = '\n'

            fun fromString(value: String) =
                PendingMemberships(value.split(SEPARATOR).toSet())
        }

        override fun toString() = uids.joinToString(SEPARATOR.toString())

    }


    // factory

    object Factory: AndroidGroupFactory<LocalGroup> {
        override fun fromProvider(addressBook: AndroidAddressBook<out AndroidContact, LocalGroup>, values: ContentValues) =
                LocalGroup(addressBook, values)
    }

}